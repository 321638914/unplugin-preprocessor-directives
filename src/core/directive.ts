import type { Directive, DirectiveContext } from '../types'

export function defineDirective(
  directive: Directive | ((ctx: DirectiveContext) => Directive),
) {
  return directive
}
