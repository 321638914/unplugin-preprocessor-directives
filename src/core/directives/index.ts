import IfDirective from './if'
import DefineDirective from './define'
import ErrorDirective from './error'

export const builtinDirectives = [DefineDirective, IfDirective, ErrorDirective]
