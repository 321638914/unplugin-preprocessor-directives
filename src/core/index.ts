export * from './context'
export * from './directive'
export * from './directives'
export * from './unplugin'
