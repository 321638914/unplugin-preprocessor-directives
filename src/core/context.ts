import process from 'node:process'
import { loadEnv } from 'vite'
import XRegExp from 'xregexp'
import type {
  Directive,
  MatchGroup,
  RecursiveDirective,
  UserOptions,
} from '../types'

import { builtinDirectives } from './directive'

export class Context {
  cwd: string
  XRegExp: typeof XRegExp
  env: Record<string, string>
  directives: Directive[] = []
  constructor({ cwd = process.cwd(), directives = [] }: UserOptions = {}) {
    this.cwd = cwd
    this.XRegExp = XRegExp
    this.env = loadEnv(
      process.env.NODE_ENV || 'development',
      process.cwd(),
      '',
    )

    this.directives = [...directives, ...builtinDirectives].map(v =>
      typeof v === 'function' ? v(this) : v,
    )
  }

  replace() { }

  replaceRecursive({
    code,
    id,
    directive,
  }: {
    code: string
    id: string
    directive: RecursiveDirective
  }) {
    const startRegex = new RegExp(directive.pattern.start, 'mi')
    const endRegex = new RegExp(directive.pattern.end, 'mi')

    const replace = (code: string) => {
      const matches = XRegExp.matchRecursive(
        code,
        startRegex.source,
        endRegex.source,
        'gmi',
        {
          valueNames: ['between', 'left', 'match', 'right'],
        },
      )
      let builder = ''
      const matchGroup: MatchGroup = {
        left: null,
        match: '',
        right: null,
      }

      if (!matches.length)
        builder += code

      matches.forEach((match) => {
        switch (match.name) {
          case 'between':
            builder += match.value
            break
          case 'left':
            matchGroup.left = startRegex.exec(match.value)
            break
          case 'match':
            matchGroup.match = match.value
            break
          case 'right':
            matchGroup.right = endRegex.exec(match.value)
            builder += directive.processor({
              id,
              code,
              directive,
              ctx: this,

              replace,
              matchGroup,
            })
            break
        }
      })

      return builder
    }

    return replace(code)
  }

  transform(code: string, id: string) {
    const data = this.directives.reduce((acc, directive) => {
      acc = directive.nested
        ? this.replaceRecursive({
          code,
          id,
          directive: directive as RecursiveDirective,
        })
        : acc
      return acc
    }, code)

    // if (id.includes('main.ts')) {
    //   console.log(data);
    // }

    return data
  }
}
