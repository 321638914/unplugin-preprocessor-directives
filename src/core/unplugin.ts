import type { UnpluginFactory } from 'unplugin'
import { createUnplugin } from 'unplugin'
import type { UserOptions } from '../types'
import { Context } from './context'

export const unpluginFactory: UnpluginFactory<UserOptions | undefined> = (
  options,
) => {
  const ctx = new Context(options)
  return {
    name: 'unplugin-preprocessor-directives',
    enforce: 'pre',
    transform: (code, id) => ctx.transform(code, id),
    vite: {
      configResolved(config) {
        ctx.env = { ...ctx.env, ...config.env }
      },
    },
  }
}

export const unplugin = /* #__PURE__ */ createUnplugin(unpluginFactory)

export default unplugin
