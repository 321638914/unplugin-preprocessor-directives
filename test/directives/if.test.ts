import { resolve } from 'node:path'
import fs from 'node:fs'
import { describe, expect, it } from 'vitest'
import { Context } from '../../src/core/context'

describe('css', () => {
  const root = resolve(__dirname, '../fixtures/css')
  const ctx = new Context({
    cwd: root,
  })
  function transform(fileName: string) {
    const filePath = resolve(root, fileName)
    const code = fs.readFileSync(filePath, { encoding: 'utf-8' })
    return ctx.transform(code, filePath)?.code
  }

  it('if', () => {
    expect(transform('if.css')).toMatchSnapshot()
  })

  it('else', () => {
    expect(transform('else.css')).toMatchSnapshot()
  })

  it('elif', () => {
    expect(transform('elif.css')).toMatchSnapshot()
  })

  it('nested', () => {
    expect(transform('nested.css')).toMatchSnapshot()
  })

  it('empty', () => {
    expect(transform('empty.css')).toBe(undefined)
  })
})

describe('js', () => {
  const root = resolve(__dirname, '../fixtures/js')
  const ctx = new Context({
    cwd: root,
  })
  function transform(fileName: string) {
    const filePath = resolve(root, fileName)
    const code = fs.readFileSync(filePath, { encoding: 'utf-8' })
    return ctx.transform(code, filePath)?.code
  }

  it('if', () => {
    expect(transform('if.js')).toMatchSnapshot()
  })

  it('else', () => {
    expect(transform('else.js')).toMatchSnapshot()
  })

  it('elif', () => {
    expect(transform('elif.js')).toMatchSnapshot()
  })

  it('nested', () => {
    expect(transform('nested.js')).toMatchSnapshot()
  })

  it('empty', () => {
    expect(transform('empty.js')).toBe(undefined)
  })
})
describe('html', () => {
  const root = resolve(__dirname, '../fixtures/html')
  const ctx = new Context({
    cwd: root,
  })
  function transform(fileName: string) {
    const filePath = resolve(root, fileName)
    const code = fs.readFileSync(filePath, { encoding: 'utf-8' })
    return ctx.transform(code, filePath)?.code
  }

  it('if', () => {
    expect(transform('if.html')).toMatchSnapshot()
  })

  it('else', () => {
    expect(transform('else.html')).toMatchSnapshot()
  })

  it('elif', () => {
    expect(transform('elif.html')).toMatchSnapshot()
  })

  it('nested', () => {
    expect(transform('nested.html')).toMatchSnapshot()
  })

  it('empty', () => {
    expect(transform('empty.html')).toBe(undefined)
  })
})
