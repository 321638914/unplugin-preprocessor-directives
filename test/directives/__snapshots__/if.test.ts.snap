// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`css > elif 1`] = `
"body {
content: \\"TEST\\";
}
"
`;

exports[`css > else 1`] = `
"body {
content: \\"!DEV else\\";
}
"
`;

exports[`css > if 1`] = `
"body {
content: \\"DEV\\";
}
"
`;

exports[`css > nested 1`] = `
"body {
content: \\"else\\";
}
"
`;

exports[`html > elif 1`] = `
"<div>TEST</div>
"
`;

exports[`html > else 1`] = `
"<div>!DEV else</div>
"
`;

exports[`html > if 1`] = `
"<div>DEV</div>
"
`;

exports[`html > nested 1`] = `
"<div>
<div>else</div>
</div>
"
`;

exports[`js > elif 1`] = `
"console.log(\\"TEST\\")
"
`;

exports[`js > else 1`] = `
"console.log(\\"!DEV else\\")
"
`;

exports[`js > if 1`] = `
"console.log(\\"DEV\\")
"
`;

exports[`js > nested 1`] = `
"console.log(\\"else\\")
"
`;
